Delete Repository in Drone DB

ulrichSchreiner opened the pull request:
Implement the deletion of a Repository in Drone. Also remove the Webhook and the deploy key from the remote vcs. Gitlab is implemented/tested, github untested. Bitbucket and gogs are empty.

bradrydzewski commented on plugin/remote/gogs/gogs.go:
```diff
@@ -153,6 +153,12 @@ func (r *Gogs) Activate(user *model.User, repo *model.Repo, link string) error {
+// Deactivate removes a repository by removing all the post-commit hooks
+// which are equal to link and removing the SSH deploy key.
+func (r *Gogs) Deactivate(user *model.User, repo *model.Repo, link string) error {
+	return fmt.Errorf("Remove %#v in gots not implemented", *repo)
```
minor spelling. Gogs

bradrydzewski commented on server/handler/repo.go:
```diff
@@ -52,6 +52,42 @@ func DelRepo(c web.C, w http.ResponseWriter, r *http.Request) {
+		var hook = fmt.Sprintf("%s/api/hook/%s/%s", httputil.GetURL(r), repo.Remote, repo.Token)
+		if err := remote.Deactivate(user, repo, hook); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
```
I think we should proceed with deleting the repository from Drone even if it fails to delete the key or hook. We could log the error for debugging purposes, but it would proceed to the delete step.

The reason is that we cannot always control what is happening in GitHub, etc. There are edge cases I can think of where I want to delete the repository from Drone (perhaps as the Drone admin) but I don't have access to the repository in GitHub to remove the hooks.

Bugagazavr replied:
Yep need to compare errors, and return multiple errors http://godoc.org/github.com/augustoroman/multierror

ulrichSchreiner replied:
as far as i understand brad's note it would be better to log the error's from the remote (github, gitlab, ...), but not return the error to the caller. so i don't think his suggestion is to accumulate them but to log some of them and don't return with an error.

bradrydzewski commented:
Added my notes. @Bugagazavr would you be able to share your thoughts on improving this REST endpoint to better conform to principles?

Bugagazavr commented:
i think wee need to separate deactivation ( `PATCH` )  and removing method ( `DELETE` )

ulrichSchreiner commented:
Please note that the semantic of the REST-API now changed. In the current master the `DELETE /api/repos/:host/:owner/:name` only deactivated it, now it really deletes the repository. I don't know if any systems are currently calling this endpoint.

bradrydzewski commented:
@Bugagazavr I think we need to be a bit more explicit with the API endpoint. The PATCH endpoint, doesn't really convey the purpose of the endpoint. It is not just updating fields, it is suspending all further builds for your repository. This feels like it should be a bit more explicit.

I researched this a bit and found many REST APIs (Google, Digital Ocean, Stripe, etc) put actions in the URL. I wasn't really a fan of this approach, but after further consideration, I think the fact that it is self-documenting is pretty important. For example:

```
POST https://api.stripe.com/v1/charges/{CHARGE_ID}/dispute/close
```

So perhaps we do this instead:

```
POST /api/repos/{host}/{owner}/{name}/deactivate
```

ulrichSchreiner commented:
I would vote for @bradrydzewski proposal with the `POST ..../deactivate` call. It is more intuitiv than a `PATCH`.

Bugagazavr commented:
maybe?

`POST /api/repos/{host}/{owner}/{name}/activate`
`DELETE /api/repos/{host}/{owner}/{name}/deactivate`

ulrichSchreiner commented:
hm,

in this case we would have 
```
POST /api/repos/{host}/{owner}/{name}/activate
DELETE /api/repos/{host}/{owner}/{name}/deactivate
DELETE /api/repos/{host}/{owner}/{name}
```
but the
```
DELETE /api/repos/{host}/{owner}/{name}/deactivate
```
does not deletey anything, it only deactivates. so the `DELETE` is a little misleading.

daMupfel commented:
and GET /api/repos/{host}/{owner}/{name} is getting new field to show if activated or not? Or how do you want to query the active status?
I would like to have a common verb, not to. For example enabled and denn POST and DELETE or something like that.

ulrichSchreiner commented:
I think the current `GET` returns a JSON struct with the `Active` field.

bradrydzewski commented:
Iet's modify this PR to the following:

```
POST /api/repos/{host}/{owner}/{name}/deactivate
DELETE /api/repos/{host}/{owner}/{name}
```

we could also modify the activation endpoint (below) however, depending on #776 it may not be necessary.

```
POST /api/repos/{host}/{owner}/{name}/activate
```

ulrichSchreiner commented:
ok, i changed the `deactivate` endpoint to `POST` and i changed it to return the repo-data as json, because it is some sort of "update" so i think it would be great to return the new data to the client. the activation endpoint is unchanged.

i will do a PR on the docs-repo to reflect this changes.

bradrydzewski commented on server/handler/repo.go:
```diff
@@ -7,6 +7,7 @@ import (
 
 	"github.com/drone/drone/plugin/remote"
 	"github.com/drone/drone/server/datastore"
+	"github.com/drone/drone/shared/build/log"
```
I realize this is confusing (sorry about that) but the build log only exists for the CLI and is going to be deprecated for something smarter. Let's use the standard `log` package instead

bradrydzewski commented on client/repos.go:
```diff
@@ -35,6 +35,12 @@ func (s *RepoService) Enable(host, owner, name string) error {
 // DELETE /api/repos/{host}/{owner}/{name}
 func (s *RepoService) Disable(host, owner, name string) error {
 	var path = fmt.Sprintf("/api/repos/%s/%s/%s", host, owner, name)
```
update method and path?

ulrichSchreiner commented:
ok, no using standard log package. there was an error too when formatting the error message. i use the prfix `[ERROR]` so if you like you can use [hashicorp logutils](https://github.com/hashicorp/logutils) which is a simple but great tool around the default log package.

bradrydzewski commented:
Thanks! I definitely like the idea of outputting log levels and using a more sophisticated logging package, and will play around with the hashicorp implementation. I've also been looking at [logrus](github.com/Sirupsen/logrus) due to its support for custom log formatters.

Note: one minor change required to `client/repos.go` to change from `PATCH` to `POST` and update path to `/deactive` as well :)

bradrydzewski commented:
merged. thanks for your patience! :)
